File: packages/core/android/src/main/java/com/posedetection/engine/Geometry.kt
package com.posedetection.engine

import com.posedetection.Skeleton
import kotlin.math.abs
import kotlin.math.acos
import kotlin.math.sqrt

/** Pure functions over the flat landmark buffer. No allocation, no state, no camera. */
internal object Geometry {
    /**
     * The angle at `vertex`, in degrees, 0 to 180.
     *
     * MediaPipe divides x by width and y by height, so on a non-square frame the normalized space
     * is anisotropic and an angle read straight off it is wrong by tens of degrees. The frame size
     * is what puts both axes back in a common unit.
     *
     * `Float.NaN` when the triangle is degenerate: 0 would be indistinguishable from a folded
     * joint.
     */
    fun angleDegrees(
        landmarks: FloatArray,
        proximal: Int,
        vertex: Int,
        distal: Int,
        frameWidth: Int,
        frameHeight: Int,
    ): Float {
        if (frameWidth <= 0 || frameHeight <= 0) return Float.NaN
        val aspect = frameWidth.toFloat() / frameHeight.toFloat()

        val vx = landmarks[vertex * Skeleton.LANDMARK_STRIDE]
        val vy = landmarks[vertex * Skeleton.LANDMARK_STRIDE + 1]

        val ax = (landmarks[proximal * Skeleton.LANDMARK_STRIDE] - vx) * aspect
        val ay = landmarks[proximal * Skeleton.LANDMARK_STRIDE + 1] - vy
        val bx = (landmarks[distal * Skeleton.LANDMARK_STRIDE] - vx) * aspect
        val by = landmarks[distal * Skeleton.LANDMARK_STRIDE + 1] - vy

        val magnitude = sqrt((ax * ax + ay * ay) * (bx * bx + by * by))
        if (magnitude < EPSILON) return Float.NaN

        // Floating point can push this a hair outside [-1, 1], where acos returns NaN.
        val cosine = ((ax * bx + ay * by) / magnitude).coerceIn(-1f, 1f)
        return Math.toDegrees(acos(cosine).toDouble()).toFloat()
    }

    /**
     * Direction of the angle's bisector, for placing the arc and its label. Takes projected screen
     * pixels: a direction taken before projection lands outside the joint on a mirrored preview.
     */
    fun bisectorRadians(
        proximalX: Float,
        proximalY: Float,
        vertexX: Float,
        vertexY: Float,
        distalX: Float,
        distalY: Float,
    ): Float {
        val ax = proximalX - vertexX
        val ay = proximalY - vertexY
        val bx = distalX - vertexX
        val by = distalY - vertexY

        val aLength = sqrt(ax * ax + ay * ay)
        val bLength = sqrt(bx * bx + by * by)
        if (aLength < EPSILON || bLength < EPSILON) return Float.NaN

        val sumX = ax / aLength + bx / bLength
        val sumY = ay / aLength + by / bLength
        if (abs(sumX) < EPSILON && abs(sumY) < EPSILON) return Float.NaN

        return kotlin.math.atan2(sumY, sumX)
    }

    fun visibility(
        landmarks: FloatArray,
        joint: Int,
    ): Float = landmarks[joint * Skeleton.LANDMARK_STRIDE + Skeleton.OFFSET_VISIBILITY]

    /**
     * Visibility-weighted center of mass, written to `out[offset]` and `out[offset + 1]`.
     *
     * Hip 0.5, ankle 0.3, knee 0.2, each side carrying half of its pair's weight and scaled by its
     * own visibility, so one occluded leg shifts the result toward the leg that is actually
     * visible rather than toward the midpoint of a guess. `NaN` when nothing is visible enough to
     * weigh: a fallback would be a position the body is not in.
     *
     * Normalized frame coordinates, uncorrected. It is compared against other normalized
     * positions, which are anisotropic in the same way, and correcting one side of that
     * comparison is what would make it wrong.
     */
    fun centerOfMass(
        landmarks: FloatArray,
        out: FloatArray,
        offset: Int,
    ) {
        var x = 0f
        var y = 0f
        var total = 0f

        for (index in COM_JOINTS.indices) {
            val joint = COM_JOINTS[index]
            val base = joint * Skeleton.LANDMARK_STRIDE
            val weight = COM_WEIGHTS[index] * landmarks[base + Skeleton.OFFSET_VISIBILITY]
            if (weight <= 0f) continue
            x += landmarks[base] * weight
            y += landmarks[base + 1] * weight
            total += weight
        }

        if (total < EPSILON) {
            out[offset] = Float.NaN
            out[offset + 1] = Float.NaN
            return
        }
        out[offset] = x / total
        out[offset + 1] = y / total
    }

    /**
     * Shoulder midpoint to ankle midpoint, in normalized units and uncorrected for the same reason
     * as [centerOfMass]: it exists to be divided into other normalized distances.
     */
    fun bodySpan(landmarks: FloatArray): Float {
        val shoulderX = midpoint(landmarks, Skeleton.LEFT_SHOULDER, Skeleton.RIGHT_SHOULDER, 0)
        val shoulderY = midpoint(landmarks, Skeleton.LEFT_SHOULDER, Skeleton.RIGHT_SHOULDER, 1)
        val ankleX = midpoint(landmarks, Skeleton.LEFT_ANKLE, Skeleton.RIGHT_ANKLE, 0)
        val ankleY = midpoint(landmarks, Skeleton.LEFT_ANKLE, Skeleton.RIGHT_ANKLE, 1)

        val dx = shoulderX - ankleX
        val dy = shoulderY - ankleY
        return sqrt(dx * dx + dy * dy)
    }

    private fun midpoint(
        landmarks: FloatArray,
        left: Int,
        right: Int,
        axis: Int,
    ): Float =
        (landmarks[left * Skeleton.LANDMARK_STRIDE + axis] + landmarks[right * Skeleton.LANDMARK_STRIDE + axis]) / 2f

    private val COM_JOINTS =
        intArrayOf(
            Skeleton.LEFT_HIP,
            Skeleton.RIGHT_HIP,
            Skeleton.LEFT_KNEE,
            Skeleton.RIGHT_KNEE,
            Skeleton.LEFT_ANKLE,
            Skeleton.RIGHT_ANKLE,
        )

    private val COM_WEIGHTS = floatArrayOf(0.25f, 0.25f, 0.1f, 0.1f, 0.15f, 0.15f)

    private const val EPSILON = 1e-6f
}

File: packages/core/ios/engine/Geometry.swift
import Foundation

/// Pure functions over the flat landmark buffer. No allocation, no state, no camera.
enum Geometry {
  private static let epsilon: Float = 1e-6
  private static let degreesPerRadian = Float(180.0 / Double.pi)

  /**
   The angle at `vertex`, in degrees, 0 to 180.

   MediaPipe divides x by width and y by height, so on a non-square frame the normalized space is
   anisotropic and an angle read straight off it is wrong by tens of degrees. The frame size is
   what puts both axes back in a common unit.

   `Float.nan` when the triangle is degenerate: 0 would be indistinguishable from a folded joint.
   */
  static func angleDegrees(
    _ landmarks: [Float],
    proximal: Int,
    vertex: Int,
    distal: Int,
    frameWidth: Int,
    frameHeight: Int
  ) -> Float {
    guard frameWidth > 0, frameHeight > 0 else { return .nan }
    let aspect = Float(frameWidth) / Float(frameHeight)

    let vx = landmarks[vertex * Skeleton.landmarkStride]
    let vy = landmarks[vertex * Skeleton.landmarkStride + 1]

    let ax = (landmarks[proximal * Skeleton.landmarkStride] - vx) * aspect
    let ay = landmarks[proximal * Skeleton.landmarkStride + 1] - vy
    let bx = (landmarks[distal * Skeleton.landmarkStride] - vx) * aspect
    let by = landmarks[distal * Skeleton.landmarkStride + 1] - vy

    let magnitude = ((ax * ax + ay * ay) * (bx * bx + by * by)).squareRoot()
    if magnitude < epsilon { return .nan }

    // Floating point can push this a hair outside [-1, 1], where acos returns NaN.
    let cosine = min(max((ax * bx + ay * by) / magnitude, -1), 1)
    return acos(cosine) * degreesPerRadian
  }

  /**
   Direction of the angle's bisector, for placing the arc and its label. Takes projected screen
   pixels: a direction taken before projection lands outside the joint on a mirrored preview.
   */
  static func bisectorRadians(
    proximalX: Float,
    proximalY: Float,
    vertexX: Float,
    vertexY: Float,
    distalX: Float,
    distalY: Float
  ) -> Float {
    let ax = proximalX - vertexX
    let ay = proximalY - vertexY
    let bx = distalX - vertexX
    let by = distalY - vertexY

    let aLength = (ax * ax + ay * ay).squareRoot()
    let bLength = (bx * bx + by * by).squareRoot()
    if aLength < epsilon || bLength < epsilon { return .nan }

    let sumX = ax / aLength + bx / bLength
    let sumY = ay / aLength + by / bLength
    if abs(sumX) < epsilon && abs(sumY) < epsilon { return .nan }

    return atan2(sumY, sumX)
  }

  static func visibility(_ landmarks: [Float], joint: Int) -> Float {
    return landmarks[joint * Skeleton.landmarkStride + Skeleton.offsetVisibility]
  }

  /**
   Visibility-weighted center of mass, written to `out[offset]` and `out[offset + 1]`.

   Hip 0.5, ankle 0.3, knee 0.2, each side carrying half of its pair's weight and scaled by its
   own visibility, so one occluded leg shifts the result toward the leg that is actually visible
   rather than toward the midpoint of a guess. `NaN` when nothing is visible enough to weigh: a
   fallback would be a position the body is not in.

   Normalized frame coordinates, uncorrected. It is compared against other normalized positions,
   which are anisotropic in the same way, and correcting one side of that comparison is what would
   make it wrong.
   */
  static func centerOfMass(_ landmarks: [Float], into out: inout [Float], at offset: Int) {
    var sumX: Float = 0
    var sumY: Float = 0
    var total: Float = 0

    for index in comJoints.indices {
      let base = comJoints[index] * Skeleton.landmarkStride
      let weight = comWeights[index] * landmarks[base + Skeleton.offsetVisibility]
      if weight <= 0 { continue }
      sumX += landmarks[base] * weight
      sumY += landmarks[base + 1] * weight
      total += weight
    }

    if total < epsilon {
      out[offset] = .nan
      out[offset + 1] = .nan
      return
    }
    out[offset] = sumX / total
    out[offset + 1] = sumY / total
  }

  /**
   Shoulder midpoint to ankle midpoint, in normalized units and uncorrected for the same reason as
   `centerOfMass`: it exists to be divided into other normalized distances.
   */
  static func bodySpan(_ landmarks: [Float]) -> Float {
    let shoulderX = midpoint(landmarks, Skeleton.leftShoulder, Skeleton.rightShoulder, axis: 0)
    let shoulderY = midpoint(landmarks, Skeleton.leftShoulder, Skeleton.rightShoulder, axis: 1)
    let ankleX = midpoint(landmarks, Skeleton.leftAnkle, Skeleton.rightAnkle, axis: 0)
    let ankleY = midpoint(landmarks, Skeleton.leftAnkle, Skeleton.rightAnkle, axis: 1)

    let dx = shoulderX - ankleX
    let dy = shoulderY - ankleY
    return (dx * dx + dy * dy).squareRoot()
  }

  private static func midpoint(_ landmarks: [Float], _ left: Int, _ right: Int, axis: Int) -> Float {
    let lhs = landmarks[left * Skeleton.landmarkStride + axis]
    let rhs = landmarks[right * Skeleton.landmarkStride + axis]
    return (lhs + rhs) / 2
  }

  private static let comJoints = [
    Skeleton.leftHip, Skeleton.rightHip,
    Skeleton.leftKnee, Skeleton.rightKnee,
    Skeleton.leftAnkle, Skeleton.rightAnkle
  ]

  private static let comWeights: [Float] = [0.25, 0.25, 0.1, 0.1, 0.15, 0.15]
}
